### ISSUE TITLE
Idp AuthnStatement

### ISSUE THREAD

**Issue by simon-jouet** [tngan/samlify#349]
Hi,

First off thanks very much for the library!

So I just started today having a look at samlify I got some basic things working but I'm struggling a bit to get the full Idp flow working. I'm using https://samltest.id/ currently to test the implementation.

For the time being the IdP is fairly basic (I followed https://github.com/tngan/samlify/blob/f2b6a2f8c36dc0ff887d0442c48cd0f2c0a4a778/examples/idp/routes/sso.js)

I can get the redirect flow working, from the SP to the IdP and back to the SP but I'm getting an error "A valid authentication statement was not found in the incoming message." which looks to be because my response doesn't include any `AuthnStatement` (which it doesn't). Is there anything to do for `AuthnStatement` to be present? Looking at `binding-post.ts` it looks like it's always set to an empty string but I might have missed something.

Also a secondary question: Is there any documentation on how to specify additional attributes? The SP I must support requires attributes for the display name and a unique id which I do have but I'm not sure how to add them to the login response.

I've added some info below which hopefully might provide some context:

The Idp Metadata
```xml
<EntityDescriptor entityID="http://0cd96cec.ngrok.io/" xmlns="urn:oasis:names:tc:SAML:2.0:metadata">
  <IDPSSODescriptor WantAuthnRequestsSigned="false" protocolSupportEnumeration="urn:oasis:names:tc:SAML:2.0:protocol">
    <KeyDescriptor use="signing">
      <KeyInfo xmlns="http://www.w3.org/2000/09/xmldsig#">
        <X509Data>
          <X509Certificate><cert?</X509Certificate>
        </X509Data>
      </KeyInfo>
    </KeyDescriptor>

    <SingleSignOnService Binding="urn:oasis:names:tc:SAML:2.0:bindings:HTTP-Redirect" Location="http://0cd96cec.ngrok.io/api/v1/auth/saml/login"/>
    <SingleSignOnService Binding="urn:oasis:names:tc:SAML:2.0:bindings:HTTP-Redirect" Location="http://0cd96cec.ngrok.io/api/v1/auth/saml/logout"/>
  </IDPSSODescriptor>
</EntityDescriptor>
```

The IdP configuration
```js
const idp = IdentityProvider({
  privateKey: fs.readFileSync('./saml/privkey.pem'),
  isAssertionEncrypted: false,

  metadata: fs.readFileSync('./saml/idp.xml'),

  nameIDFormat: ['emailAddress']
});

const sp = ServiceProvider({
  metadata: fs.readFileSync('./saml/sp.xml'),
});

'GET /api/v1/auth/saml/metadata': (_req: Request, res: Response, _next: NextFunction) => {

  res.header('Content-Type', 'text/xml');
  res.send(idp.getMetadata());
},

'GET /api/v1/auth/saml/login': (req: Request, res: Response, _next: NextFunction) => {
  idp
    .parseLoginRequest(sp, 'redirect', req)
    .then(parseResult => {
      const user = {
        email: 'test@test.com',
        firstname: 'test',
        lastname: 'user',
        login: 'potato'
      };

      return idp.createLoginResponse(sp, parseResult, 'post', user);
    })
    .then(response => {
      console.log(response);

      res.render('saml/actions', response);
    });
```

Sample Response from the IdP
```xml
<samlp:Response xmlns:samlp="urn:oasis:names:tc:SAML:2.0:protocol" 
    xmlns:saml="urn:oasis:names:tc:SAML:2.0:assertion" ID="_5f0a3726-ef92-466b-9e2e-b498631f0ab5" Version="2.0" IssueInstant="2020-03-17T15:06:40.322Z" Destination="https://samltest.id/Shibboleth.sso/SAML2/POST" InResponseTo="_23e6dbffa6f7ce71c5a78344bbc67702">
    <saml:Issuer>http://0cd96cec.ngrok.io/</saml:Issuer>
    <ds:Signature xmlns:ds="http://www.w3.org/2000/09/xmldsig#">
        <ds:SignedInfo>
            <ds:CanonicalizationMethod Algorithm="http://www.w3.org/2001/10/xml-exc-c14n#"/>
            <ds:SignatureMethod Algorithm="http://www.w3.org/2001/04/xmldsig-more#rsa-sha256"/>
            <ds:Reference URI="#_5f0a3726-ef92-466b-9e2e-b498631f0ab5">
                <ds:Transforms>
                    <ds:Transform Algorithm="http://www.w3.org/2000/09/xmldsig#enveloped-signature"/>
                    <ds:Transform Algorithm="http://www.w3.org/2001/10/xml-exc-c14n#"/>
                </ds:Transforms>
                <ds:DigestMethod Algorithm="http://www.w3.org/2001/04/xmlenc#sha256"/>
                <ds:DigestValue>UYN0bzmMi8zII8Gyd22QoWf9txz01TAsDgwAlwFvZ4Q=</ds:DigestValue>
            </ds:Reference>
        </ds:SignedInfo>
        <ds:SignatureValue>g6i6VrrZCppzuTklzQUxkZ8c03DDwk6FnWfA8iAV1cd2y7Oo743S0D78exEneclOoJnqeIlslUisnJ7EAHUuESIlPMURUB19z6reaZS3TLsWHTx/cnhrpj3Lly68wZobAjcEMY4m3ebSTQ3JgkVh8KmsHeHYUUsYk6t+9O1sCGPPhaCtKovBFKf+GsxfjV3CIXEe6hLYruz97/kK+ExgmuApP/hRKLfjNSjambzRy70qkAgbNO1NQCj7B3zmupXzpoZlG5s6M1D/1JSw+IsR5QgIi2LaRcb6K0X0eUKYBYHT+5UdTk29ApLXUCDoUT+csXPXkSh3Tvc0UtBkFBFWfFm0laUxCT0DFCQrisSPvPv/5VRsJrFG5Mxweup8jEfCtCwsS+su+pKW45U8n8VZ5rOxf3OSzuzfb+FHwe64xnDGZWMZ6pp6ZEtBkKxRUEIttPZ+RwOQ2yF5M5g3/ZPjNo38aaXQHqzDOFGhn/yqk/vQT7ONoSuAZ9sVPiryL0CzC1439CedDSU8Eowt9GfqtDOAyAmc+ItyRBjkBJ/ffqx22dC3zztMEDzSxqZZ+02JHXZG3PjzMQOjyNfHYKpTK+kWcCle4rQ9gKNZkuNeaYDa/lMOzJhbkxyMP/33JqMkf/3db97Fi6ue6sbOUDbFMSLosKFMLqvaz4BRfAhUsP0=</ds:SignatureValue>
        <ds:KeyInfo>
            <ds:X509Data>
                <ds:X509Certificate>cert</ds:X509Certificate>
            </ds:X509Data>
        </ds:KeyInfo>
    </ds:Signature>
    <samlp:Status>
        <samlp:StatusCode Value="urn:oasis:names:tc:SAML:2.0:status:Success"/>
    </samlp:Status>
    <saml:Assertion xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" 
        xmlns:xs="http://www.w3.org/2001/XMLSchema" 
        xmlns:saml="urn:oasis:names:tc:SAML:2.0:assertion" ID="_0b7593cd-334f-4d2a-9b92-783b342fd868" Version="2.0" IssueInstant="2020-03-17T15:06:40.322Z">
        <saml:Issuer>http://0cd96cec.ngrok.io/</saml:Issuer>
        <saml:Subject>
            <saml:NameID Format="emailAddress">test@test.com</saml:NameID>
            <saml:SubjectConfirmation Method="urn:oasis:names:tc:SAML:2.0:cm:bearer">
                <saml:SubjectConfirmationData NotOnOrAfter="2020-03-17T15:11:40.322Z" Recipient="https://samltest.id/Shibboleth.sso/SAML2/POST" InResponseTo="_23e6dbffa6f7ce71c5a78344bbc67702"/>
            </saml:SubjectConfirmation>
        </saml:Subject>
        <saml:Conditions NotBefore="2020-03-17T15:06:40.322Z" NotOnOrAfter="2020-03-17T15:11:40.322Z">
            <saml:AudienceRestriction>
                <saml:Audience>https://samltest.id/saml/sp</saml:Audience>
            </saml:AudienceRestriction>
        </saml:Conditions>
    </saml:Assertion>
</samlp:Response>
```

The logs from samltest

```
2020-03-17 15:24:30 DEBUG OpenSAML.MessageDecoder.SAML2POST [3183] [default]: decoded SAML message:
<samlp:Response xmlns:samlp="urn:oasis:names:tc:SAML:2.0:protocol" xmlns:saml="urn:oasis:names:tc:SAML:2.0:assertion" ID="_65abdbd4-1494-4383-9fea-a3e07a3db15d" Version="2.0" IssueInstant="2020-03-17T15:24:30.366Z" Destination="https://samltest.id/Shibboleth.sso/SAML2/POST" InResponseTo="_1e420862fd08bf1e54b495c4911cf85b"><saml:Issuer>http://0cd96cec.ngrok.io/</saml:Issuer><ds:Signature xmlns:ds="http://www.w3.org/2000/09/xmldsig#"><ds:SignedInfo><ds:CanonicalizationMethod Algorithm="http://www.w3.org/2001/10/xml-exc-c14n#"/><ds:SignatureMethod Algorithm="http://www.w3.org/2001/04/xmldsig-more#rsa-sha256"/><ds:Reference URI="#_65abdbd4-1494-4383-9fea-a3e07a3db15d"><ds:Transforms><ds:Transform Algorithm="http://www.w3.org/2000/09/xmldsig#enveloped-signature"/><ds:Transform Algorithm="http://www.w3.org/2001/10/xml-exc-c14n#"/></ds:Transforms><ds:DigestMethod Algorithm="http://www.w3.org/2001/04/xmlenc#sha256"/><ds:DigestValue>9ix+M1tFr4G7H2Gx4Fkx7f8Bwt2/ekGl3k+EqByDMd0=</ds:DigestValue></ds:Reference></ds:SignedInfo><ds:SignatureValue>signature</ds:SignatureValue><ds:KeyInfo><ds:X509Data><ds:X509Certificate>cert</ds:X509Certificate></ds:X509Data></ds:KeyInfo></ds:Signature><samlp:Status><samlp:StatusCode Value="urn:oasis:names:tc:SAML:2.0:status:Success"/></samlp:Status><saml:Assertion xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xs="http://www.w3.org/2001/XMLSchema" xmlns:saml="urn:oasis:names:tc:SAML:2.0:assertion" ID="_f5e1b546-bd42-456e-a0c1-e2cc6b4f8f71" Version="2.0" IssueInstant="2020-03-17T15:24:30.366Z"><saml:Issuer>http://0cd96cec.ngrok.io/</saml:Issuer><saml:Subject><saml:NameID Format="emailAddress">test@test.com</saml:NameID><saml:SubjectConfirmation Method="urn:oasis:names:tc:SAML:2.0:cm:bearer"><saml:SubjectConfirmationData NotOnOrAfter="2020-03-17T15:29:30.366Z" Recipient="https://samltest.id/Shibboleth.sso/SAML2/POST" InResponseTo="_1e420862fd08bf1e54b495c4911cf85b"/></saml:SubjectConfirmation></saml:Subject><saml:Conditions NotBefore="2020-03-17T15:24:30.366Z" NotOnOrAfter="2020-03-17T15:29:30.366Z"><saml:AudienceRestriction><saml:Audience>https://samltest.id/saml/sp</saml:Audience></saml:AudienceRestriction></saml:Conditions></saml:Assertion></samlp:Response>
2020-03-17 15:24:30 DEBUG OpenSAML.MessageDecoder.SAML2 [3183] [default]: extracting issuer from SAML 2.0 protocol message
2020-03-17 15:24:30 DEBUG OpenSAML.MessageDecoder.SAML2 [3183] [default]: message from (http://0cd96cec.ngrok.io/)
2020-03-17 15:24:30 DEBUG OpenSAML.MessageDecoder.SAML2 [3183] [default]: searching metadata for message issuer...
2020-03-17 15:24:30 DEBUG OpenSAML.SecurityPolicyRule.MessageFlow [3183] [default]: evaluating message flow policy (replay checking on, expiration 60)
2020-03-17 15:24:30 DEBUG XMLTooling.StorageService [3183] [default]: inserted record (_65abdbd4-1494-4383-9fea-a3e07a3db15d) in context (MessageFlow) with expiration (1584458910)
2020-03-17 15:24:30 DEBUG OpenSAML.SecurityPolicyRule.XMLSigning [3183] [default]: validating signature profile
2020-03-17 15:24:30 DEBUG OpenSAML.SecurityPolicyRule.XMLSigning [3183] [default]: signature verified against message issuer
2020-03-17 15:24:30 DEBUG Shibboleth.SSO.SAML2 [3183] [default]: processing message against SAML 2.0 SSO profile
2020-03-17 15:24:30 DEBUG Shibboleth.SSO.SAML2 [3183] [default]: extracting issuer from SAML 2.0 assertion
2020-03-17 15:24:30 DEBUG OpenSAML.SecurityPolicyRule.MessageFlow [3183] [default]: evaluating message flow policy (replay checking on, expiration 60)
2020-03-17 15:24:30 DEBUG XMLTooling.StorageService [3183] [default]: inserted record (_f5e1b546-bd42-456e-a0c1-e2cc6b4f8f71) in context (MessageFlow) with expiration (1584458910)
2020-03-17 15:24:30 DEBUG OpenSAML.SecurityPolicyRule.BearerConfirmation [3183] [default]: assertion satisfied bearer confirmation requirements
2020-03-17 15:24:30 WARN Shibboleth.SSO.SAML2 [3183] [default]: error processing incoming assertion: A valid authentication statement was not found in the incoming message.
```

Best regards,
Simon

**Comment by tngan** [comment]
@simon-jouet Hi Simon, so you are using samlify as IdP and use samltest.id to be the tested SP, right?

For more complicated response, you can customize and replace the tag during runtime. These use cases are included in the test suite.

https://github.com/tngan/samlify/blob/c6a17ff815de04779fb245a8a6e1088df7dfd513/test/flow.ts#L349-L358

https://github.com/tngan/samlify/blob/c6a17ff815de04779fb245a8a6e1088df7dfd513/test/flow.ts#L313-L324

The callback function to do runtime tag replacement.

https://github.com/tngan/samlify/blob/c6a17ff815de04779fb245a8a6e1088df7dfd513/test/flow.ts#L45-L76

**Comment by simon-jouet** [comment]
Thanks @tngan for the details, I also just got more details from the service provider that we need to support so I will get back to this soon. I will close this issue for now, I hope that with the new details and the examples provided I will be able to get something working!

**Comment by matthewsanders** [comment]
I am in the same boat and the test suite logic does very little to shed light on the solution here.  Is there any ETA on some proper documentation for creating your own Idp with this "library"?

I find the library is very cumbersome so far in this regard.  It appears that if you want to override tags it is an ALL or NOTHING affair and you can't simply rely on samlify to select reasonable defaults.

so far I had to explicitly set the template:

```
const idp = samlify.IdentityProvider({
    loginResponseTemplate: samlify.SamlLib.defaultLoginResponseTemplate,
```

why not default to this since this is what it uses when you do not override anyway?

I also had to copy and paste much of the internal logic that does the tag replacement:

```
async function postUiSamlAcs (req : Request, res : Response) : Promise<void> {
        const requestData = await idp.parseLoginRequest(sp, 'post', req);

        // this was exposed but doesn't seem to be exposed outside of the library?
        function get(obj : any, path : any, defaultValue : any) {
            return path.split('.')
                .reduce((a : any, c : any) => (a && a[c] ? a[c] : (defaultValue || null)), obj);
        }

        const extract = requestData.extract; // is this even needed for anything?
        const user = { email: 'mytest@somedomain.com' };
        const responseObj = await idp.createLoginResponse(sp, requestData, 'post', user, (template: string) : BindingContext => {
            const idpSetting = idp.entitySetting;
            const spSetting = sp.entitySetting;
            const id = (idpSetting as any).generateID();
            const metadata = {
              idp: idp.entityMeta,
              sp: sp.entityMeta,
            };
            const nameIDFormat = idpSetting.nameIDFormat;
            const selectedNameIDFormat = Array.isArray(nameIDFormat) ? nameIDFormat[0] : nameIDFormat;
            
            const base = metadata.sp.getAssertionConsumerService('post');
            const nowTime = new Date();
            const spEntityID = metadata.sp.getEntityID();
            const fiveMinutesLaterTime = new Date(nowTime.getTime());
            fiveMinutesLaterTime.setMinutes(fiveMinutesLaterTime.getMinutes() + 5);
            const tenHoursLaterTime = new Date(nowTime.getTime());
            tenHoursLaterTime.setHours(tenHoursLaterTime.getHours() + 10);
            const fiveMinutesLater = fiveMinutesLaterTime.toISOString();
            const tenHoursLater = tenHoursLaterTime.toISOString();
            const now = nowTime.toISOString();
            const acl = metadata.sp.getAssertionConsumerService('post');

            const sessionIndex = '_be9967abd904ddcae3c0eb4189adbe3f71e327cf93'; // how do we generate this?
            
            // if you elect to pass a custom callback you have to override EVERYTHING manually
            // couldn't you default the majority of the boilerplate and override what we feed to the API?
            const tvalue: any = {
                ID: id,
                AssertionID: (idpSetting as any).generateID(),
                Destination: base,
                Audience: spEntityID,
                EntityID: spEntityID,
                SubjectRecipient: acl,
                Issuer: metadata.idp.getEntityID(),
                IssueInstant: now,
                AssertionConsumerServiceURL: acl,
                StatusCode: samlify.Constants.StatusCode.Success,
                // can be customized
                ConditionsNotBefore: now,
                ConditionsNotOnOrAfter: fiveMinutesLater,
                SubjectConfirmationDataNotOnOrAfter: fiveMinutesLater,
                NameIDFormat: selectedNameIDFormat,
                NameID: user.email || '',
                InResponseTo: get(requestData, 'extract.request.id', ''),
                AuthnStatement: stripIndent`
                <saml:AuthnStatement AuthnInstant="${now}" SessionNotOnOrAfter="${tenHoursLater}" SessionIndex="${sessionIndex}">
                    <saml:AuthnContext>
                        <saml:AuthnContextClassRef>urn:oasis:names:tc:SAML:2.0:ac:classes:Password</saml:AuthnContextClassRef>
                    </saml:AuthnContext>
                </saml:AuthnStatement>
                `,
                AttributeStatement: '',
            };
            
            return {
                id : tvalue.ID,
                context : samlify.SamlLib.replaceTagsByValue(template, tvalue)
            };
        } );

        res.status(200).send(postFunc( responseObj ));
    }
```

I have no idea how correct this as it isn't working with either samltest.id or my third party vendor I'm trying to integrate with.

Any insights would be greatly appreciated.

**Comment by simon-jouet** [comment]
Hi @matthewsanders 

Looks like you have spent more time than i have to work on this. Since posting the issue I haven't had much time at all to go back to it but thanks for posting what you've got so far it might help out when I'm going back to this (which should be soon)

I agree that the current documentation for the IdP side of things is quite limited, but I also understand that as an open source project the documentation is maybe not the top priority. 

@tngan would you have some insight on some of the points raised by @matthewsanders by any chance? Would you have any sample implementation you could share with the AuthnStatement being used and/or something that is compatible samltest.id as a baseline?

Thanks very much both for your replies

**Comment by tngan** [comment]
@marcbachmann @simon-jouet 

Sorry for late reply. Yes, we are trying to keep the iDP part minimal, because we are not targeting to build an iDP at the very beginning, the purpose of building this library is to provide an easier way to integrate your service application with SAML login to iDP.

If you don't specify `loginResponseTemplate`, it will default to use `defaultLoginResponseTemplate` while building the response.

https://github.com/tngan/samlify/blob/46ce077e6278e28bb422e30f66e80abe166c622d/src/binding-post.ts#L124-L132

Whenever you need to use your own template with custom tags instead of the default one, I cannot put in the default value to your custom tags because the library won't know which tag should the default value maps to.

If you use default template, you don't need to specify the tag replacement callback, those default values will be replaced and mapped into the corresponding tags.

Or your use case is to use default template but having your own value like you want to change some of the value like expiry date of the response, that would make sense that I could also allow user to override the value for the default template, it is a valid enhancements to do.

I have removed the iDP example after migrating to v2, maybe it's a good time to add it back, I should also try samltest.id to test both SP and iDP side.

@matthewsanders  Let me know if I understand your situation correctly, thank you so much for your feedback! :)

**Comment by matthewsanders** [comment]
I actually did get this to work after stepping through the library and comparing the `rawSamlResponse` inside `base64LoginResponse` with known good response data from other providers.

I can definitely tell @tngan has great depth of knowledge of SAML 2.0 and I respect how quickly you respond.  I hope I didn't sound too harsh on my first post as I do greatly appreciate open source projects.  I really only ever work with open source software for a multitude of reasons.

I think the biggest thing with any library is to assume your audience is not a subject matter expert.  A great library should be easy to use correctly and hard to use incorrectly.  The focus here is around abstracting the nitty gritty details of SAML 2.0 so anyone could use it without needing to reference the spec.  I do realize, however, that this is a work in progress and that Idp wasn't an initial target.  I still found it hard to follow up front and I basically had to become more of an SME for SAML 2.0 to use it.

As for the actual problems I ran into.  I did not intend to use a custom template but I found to get it working it was required currently.  If I didn't set `loginResponseTemplate` explicitly in my Idp it would be null so I couldn't use the `customTagReplacement` functionality which seemed like the only way I could customize the `AuthnStatement` and `AttributeStatement` even with the default template.

I also noticed that when I was explicitly setting it with the default:
`loginResponseTemplate: samlify.SamlLib.defaultLoginResponseTemplate,`

The AttributeStatement was somehow being replaced with an empty AttributeStatement XML node.  This was tripping up validation and eventually was the final problem I had to work around.

So in the end I had to copy/paste the default template explicitly:

```
loginResponseTemplate: {
        context: stripIndent`
        <samlp:Response xmlns:samlp="urn:oasis:names:tc:SAML:2.0:protocol" xmlns:saml="urn:oasis:names:tc:SAML:2.0:assertion" ID="{ID}" Version="2.0" IssueInstant="{IssueInstant}" Destination="{Destination}" InResponseTo="{InResponseTo}">
        <saml:Issuer>{Issuer}</saml:Issuer>
        <samlp:Status>
            <samlp:StatusCode Value="{StatusCode}"/>
        </samlp:Status>
        <saml:Assertion xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xs="http://www.w3.org/2001/XMLSchema" xmlns:saml="urn:oasis:names:tc:SAML:2.0:assertion" ID="{AssertionID}" Version="2.0" IssueInstant="{IssueInstant}"><saml:Issuer>{Issuer}</saml:Issuer><saml:Subject><saml:NameID Format="{NameIDFormat}">{NameID}</saml:NameID><saml:SubjectConfirmation Method="urn:oasis:names:tc:SAML:2.0:cm:bearer"><saml:SubjectConfirmationData NotOnOrAfter="{SubjectConfirmationDataNotOnOrAfter}" Recipient="{SubjectRecipient}" InResponseTo="{InResponseTo}"/></saml:SubjectConfirmation></saml:Subject><saml:Conditions NotBefore="{ConditionsNotBefore}" NotOnOrAfter="{ConditionsNotOnOrAfter}"><saml:AudienceRestriction><saml:Audience>{Audience}</saml:Audience></saml:AudienceRestriction></saml:Conditions>{AuthnStatement}{AttributeStatement}</saml:Assertion>
        </samlp:Response>
        `
    },
```
and finally I filled in the Attribute of the copy/paste callback logic I mentioned above:

```
AttributeStatement: stripIndent`
                <saml:AttributeStatement>
                    <saml:Attribute Name="mail" NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:basic">
                        <saml:AttributeValue xsi:type="xs:string">${user.email}</saml:AttributeValue>
                    </saml:Attribute>
                </saml:AttributeStatement>
                `
```

The takeaway for me here is that the copy/paste should be unnecessary if I was simply wanting to use the default template and replace only the "tags" within it.  There very well could be a better way to use the library to avoid what I did, but that goes back to making it easy to use correctly and hard to use incorrectly.  I would dig into it more, but another party I'm working with has an Idp setup running on the platform that we are migrating our users into so I may no longer require this particular Idp after all.

From my very limited knowledge of SAML 2.0 thus far it seems like the abstractions should be pretty easy to identify.  I found the onelogin developer tools took a pretty good shot at it: https://www.samltool.com/idp_metadata.php.  Much of the example requests and responses I ran across had the same fields and typically only deviated much when it came to defining the attributes.

All in all it is very promising to see this library taking shape and I hope it benefits others in the future!

**Comment by simon-jouet** [comment]
@tngan Thanks very much for the details, I'm sure that will make things a bit clearer when I get back to that (likely next week). The reason for me to use samlify for an IdP is that from what I saw it was one of the very few library to support that. I might have missed others though but in the current landscape it seemed to me like it was one of the few libraries to provide some abstraction to provide an IdP.

Regarding the samltest.id IdP it would indeed be great to have an example of that as a reference implementation.

@matthewsanders thanks for even more info! I can relate very much on the SAML 2.0 knowledge bit, I know very little about it and it was clearly a stumbling block to get things started. I had done a SP implementation before using passport-saml but that was about it (and fairly straightforware).

Sorry to ask that but would you be able to share the code you came up with so far the IdP side of things? it will definitely help in getting (re)started!

**Comment by zag2art** [comment]
Probably that issue should go into docs. I spent a few days to get the same result as in @matthewsanders' snippet. Of course now I know a lot of saml internals but I would prefer to have a ready to use library.
